convert2rhel/unit_tests/redhatrelease_test.py

```diff
@@ -142,13 +158,30 @@ def test_yum_patch(monkeypatch, modified, caplog):
         assert "Skipping patching, yum configuration file not modified" in caplog.text
 
 
-YUM_CONF_WITHOUT_DISTROVERPKG = """[main]
-installonly_limit=3
+@unit_tests.mock(os.path, "isfile", lambda x: True)
+def test_get_os_release_filepath(caplog):
+    assert get_os_release_filepath() == "/etc/os-release"
 
-#  This is the default"""
+    os.path.isfile = lambda x: False
 
-YUM_CONF_WITH_DISTROVERPKG = """[main]
-installonly_limit=3
-distroverpkg=centos-release
+    with pytest.raises(SystemExit):
+        get_os_release_filepath()
+        assert "Error:" in caplog.records[-1].message
 
-#  This is the default"""
+
+@pytest.mark.parametrize(
+    ("filepath", "file_content", "raise_error"),
+    (("etc/os-release", "", True), ("/etc/os-release", "NAME=Test\nVERSION=1\n", False)),
+)
+def test_get_os_release_content(filepath, file_content, raise_error, monkeypatch, caplog):
+    get_os_release_filepath_mock = mock.Mock(return_value=filepath)
+    monkeypatch.setattr(redhatrelease, "get_os_release_filepath", get_os_release_filepath_mock)
+
+    if raise_error:
+        get_file_content_mock = mock.Mock(side_effect=EnvironmentError)
+        monkeypatch.setattr(redhatrelease.utils, "get_file_content", get_file_content_mock)
+        with pytest.raises(SystemExit):
+            assert get_os_release_content()
+    else:
+        monkeypatch.setattr(redhatrelease.utils, "get_file_content", mock.Mock(return_value=file_content))
+        assert "NAME=Test" in get_os_release_content()
```

Comment:
* I would definitely make the happy path a separate test from the error path.

* I'm not sure how other people feel about unit tests vs functional tests vs integration tests ... I tend to favour functional tests over unit tests as they strike a balance between testing what will occur when the code is called for real and speed of execution plus ease of debugging.  If this was recast as a functional test, I would probably use pytest's facilities to create a temporary file with my desired content and then use monkeypatch on get_os_release_filepath to return the temporary filename.  That way:
  * Except for determining the filename, I would be running the actual code that would be in production with all of the potential places where one piece calling another could go wrong.
  * I could more easily construct my tests as setting up a system environment and then testing that the code runs on it.  ie: If I wanted to test that a directory instead of a file caused the expected exception, I could replace the temporary file with a directory.  No need to mock out anything further.
  * Since I'd still be monkeypatching get_os_release_filepath() in that example, I might also have one or two additional tests that showed that get_os_release_content was doing what I expected when get_os_release_filepath
was not mocked out.  That would probably be a few error checks and one happy path test where I'd have to monkeypatch utils.get_file_content instead.

Reply:
Having separate happy tests could be good, though not a requirement since we don't do it in the rest of the codebase - so won't be a blocker

Unit tests in general is a bad term, testing as a unit doesn't mean mock everything and just hope for the best. Functional tests fit better due to the reasons you mentioned - but I'll still call these unit tests

100% anything that don't need to be mocked shouldn't. Creating temp files and mocking a path is much better than mocking a dependency. I did have some issues previously with creating a test file for some reason I don't remember right now

Reply:
I think my question for that test is no longer valid... Since the review, we found out the function had no purpose as it was not being used by anyone else, so, this test got deleted as well. But the discussion about different tests types is really good. Maybe in the future, we can think about how to improve more our tests with the help of QE folks and get something more consistent too.

Reply:
I'm on the fence here because if you write a functional test for a high level function (in terms of abstraction) you could end up mocking lots of functions many levels lower.

I think this is worth discussing on a separate meeting, to get ourselves on the same page when it comes to testing.

Reply:
Agreed. Our testes are really divergent on some files, maybe we could set up that separate meeting in the near future to discuss better on what we want to do (or we could leave it when we port c2r to leapp as well).